File: api/dovc-gateway.js
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = 5000;
const ROOT_DIR = __dirname + '/..'; // project root (one level up from api folder)

function sendJson(res, statusCode, data) {
  const payload = JSON.stringify(data);
  res.writeHead(statusCode, { 'Content-Type': 'application/json' });
  res.end(payload);
}

function loadInventory() {
  const inventoryPath = path.join(ROOT_DIR, 'inventory.json');
  try {
    return JSON.parse(fs.readFileSync(inventoryPath, 'utf-8'));
  } catch (e) {
    console.error('Failed to load inventory.json', e);
    return null;
  }
}

function handleApi(req, res) {
  if (req.method !== 'POST') {
    return sendJson(res, 405, { error: 'Method Not Allowed' });
  }
  let body = '';
  req.on('data', chunk => body += chunk);
  req.on('end', () => {
    let payload;
    try {
      payload = JSON.parse(body);
    } catch (e) {
      return sendJson(res, 400, { error: 'Invalid JSON' });
    }
    if (!payload.intent || !payload.constraints) {
      return sendJson(res, 400, { error: 'Missing intent or constraints' });
    }
    
    // Simulate latency
    setTimeout(() => {
      const inventory = loadInventory();
      if (!inventory) {
        return sendJson(res, 500, { error: 'Server error: cannot load inventory' });
      }
      
      const reqMonthly = parseInt(payload.constraints.maxMonthly, 10) || 450;
      const reqDown = parseInt(payload.constraints.maxDown, 10) || 2500;
      const maxAllowed = reqMonthly + 50;
      
      const candidates = inventory.filter(v => v.monthly_estimate <= maxAllowed);
      candidates.sort((a, b) => b.monthly_estimate - a.monthly_estimate);
      
      if (candidates.length === 0) {
        return sendJson(res, 404, { success: false, error: 'No matching vehicles found within those strict parameters.' });
      }
      
      const vehicle = candidates[0];
      const actualMonthly = Math.min(vehicle.monthly_estimate, reqMonthly - 12);
      
      const result = {
        success: true,
        matchScore: '99.8%',
        vehicle: {
          year: vehicle.year,
          make: vehicle.make,
          model: vehicle.model,
          trim: vehicle.trim,
          color: vehicle.color,
          imageUrl: vehicle.image_url,
          dealership: vehicle.dealership_name,
          address: vehicle.dealership_address
        },
        dealStructure: {
          down: `$${reqDown}`,
          monthly: `$${actualMonthly}`
        },
        translatedLog: `CARNIMBUS SECURED: Based on your strict ${payload.intent} constraint of $${reqMonthly}/mo, our systems successfully negotiated a priority allocation for a ${vehicle.year} ${vehicle.make} ${vehicle.model} ${vehicle.trim}. We matched this vehicle at ${vehicle.dealership_name} (${vehicle.zip_code} matrix) that beats your ceiling with exactly $${reqDown} down.`
      };
      
      return sendJson(res, 200, result);
    }, 1200);
  });
}

function serveStatic(filePath, res) {
  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404, { 'Content-Type': 'text/plain' });
      return res.end('Not Found');
    }
    const ext = path.extname(filePath).toLowerCase();
    const mime = {
      '.html': 'text/html',
      '.css': 'text/css',
      '.js': 'application/javascript',
      '.json': 'application/json',
      '.png': 'image/png',
      '.jpg': 'image/jpeg',
      '.svg': 'image/svg+xml'
    }[ext] || 'application/octet-stream';
    res.writeHead(200, { 'Content-Type': mime });
    res.end(data);
  });
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url, `http://${req.headers.host}`);
  if (url.pathname === '/api/dovc-gateway' || url.pathname === '/dovc-gateway') {
    return handleApi(req, res);
  }
  // Serve static files from project root
  let filePath = path.join(ROOT_DIR, url.pathname);
  if (fs.statSync(filePath).isDirectory()) {
    filePath = path.join(filePath, 'index.html');
  }
  return serveStatic(filePath, res);
});

server.listen(PORT, () => {
  console.log(`🚀 Node API server listening on http://127.0.0.1:${PORT}`);
});

File: python_api/dovc_gateway.py
import json
from pathlib import Path
from flask import Flask, request, jsonify, send_from_directory

app = Flask(__name__)

# Root workspace directory is one level up from this file's folder (python_api)
ROOT_DIR = Path.cwd()

@app.route('/')
def serve_index():
    return send_from_directory(str(ROOT_DIR), 'index.html')

@app.route('/<path:path>')
def serve_static(path):
    return send_from_directory(ROOT_DIR, path)

@app.route('/dovc-gateway', methods=['POST'])
@app.route('/api/dovc-gateway', methods=['POST'])
def dovc_gateway():
    """Mimics the original PowerShell listener and python_worker.py logic.
    Expects JSON with `intent` and `constraints`.
    """
    # Parse payload robustly
    payload = request.get_json(silent=True)
    if not payload:
        try:
            payload = json.loads(request.data.decode('utf-8'))
        except Exception:
            pass
    
    if not payload or "intent" not in payload or "constraints" not in payload:
        return jsonify({"error": "Missing intent or constraints"}), 400

    # Load inventory.json
    inventory_path = ROOT_DIR / "inventory.json"
    try:
        inventory = json.loads(inventory_path.read_text(encoding="utf-8"))
    except Exception as exc:
        app.logger.error("Failed to read inventory.json: %s", exc)
        return jsonify({"error": "Server error: cannot load inventory"}), 500

    req_monthly = int(payload['constraints'].get('maxMonthly', 450))
    req_down = int(payload['constraints'].get('maxDown', 2500))
    max_allowed = req_monthly + 50

    candidates = [v for v in inventory if v.get('monthly_estimate', 9999) <= max_allowed]
    candidates.sort(key=lambda v: v.get('monthly_estimate', 0), reverse=True)

    if not candidates:
        return jsonify({"success": False, "error": "No matching vehicles found within those strict parameters."}), 404

    vehicle = candidates[0]
    actual_monthly = min(vehicle.get('monthly_estimate', 0), req_monthly - 12)

    result = {
        "success": True,
        "matchScore": "99.8%",
        "vehicle": {
            "year": vehicle.get('year'),
            "make": vehicle.get('make'),
            "model": vehicle.get('model'),
            "trim": vehicle.get('trim'),
            "color": vehicle.get('color'),
            "imageUrl": vehicle.get('image_url'),
            "dealership": vehicle.get('dealership_name'),
            "address": vehicle.get('dealership_address')
        },
        "dealStructure": {
            "down": f"${req_down}",
            "monthly": f"${actual_monthly}"
        },
        "translatedLog": (
            f"CARNIMBUS SECURED: Based on your strict {payload['intent']} constraint of "
            f"${req_monthly}/mo, our systems successfully negotiated a priority allocation "
            f"for a {vehicle.get('year')} {vehicle.get('make')} {vehicle.get('model')} {vehicle.get('trim')}. "
            f"We matched this vehicle at {vehicle.get('dealership_name')} ({vehicle.get('zip_code')} matrix) "
            f"that beats your ceiling with exactly ${req_down} down."
        )
    }

    return jsonify(result)

@app.route('/')
def serve_index():
    return send_from_directory(str(ROOT_DIR), 'index.html')

@app.route('/<path:path>')
def serve_static(path):
    return send_from_directory(ROOT_DIR, path)

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=False)

